src/wallet.rs

```diff
@@ -0,0 +1,962 @@
+            rusk.state.clone(),
+            rusk.network,
+            self.status,
+        );
+
+        // create a state client
+        let state = State::new(rusk.state, self.status)?;
+
+        // create wallet instance
+        self.wallet = Some(WalletCore::new(self.store.clone(), state, prover));
+
+        Ok(())
+    }
+
+    /// Checks if the wallet has an active connection to the network
+    pub fn is_online(&self) -> bool {
+        self.wallet.is_some()
+    }
+
+    /// Obtain balance information for a given address
+    pub async fn get_balance(
+        &self,
+        addr: &Address,
+    ) -> Result<BalanceInfo, Error> {
+        // make sure we own this address
+        if !addr.is_owned() {
+            return Err(Error::Unauthorized);
+        }
+
+        // get balance
+        if let Some(wallet) = &self.wallet {
+            let index =
+                addr.index().expect("owned address should have an index");
+            wallet.get_balance(index).map_err(Error::from)
+        } else {
+            Err(Error::Offline)
+        }
+    }
+
+    /// Creates a new public address.
+    /// The addresses generated are deterministic across sessions.
+    pub fn new_address(&mut self) -> Address {
+        let addr = self.get_address(self.addrs.count);
+        self.addrs.count += 1;
+        addr
+    }
+
+    /// Default public address for this wallet
+    pub fn default_address(&self) -> Address {
+        let ssk = self
+            .store
+            .retrieve_ssk(0)
+            .expect("wallet seed should be available");
+        Address::new(0, ssk.public_spend_key())
+    }
+
+    fn get_address(&self, index: u64) -> Address {
+        let ssk = self
+            .store
+            .retrieve_ssk(index)
+            .expect("wallet seed should be available");
+        Address::new(index, ssk.public_spend_key())
+    }
+
+    /// Addresses that have been generated by the user
+    pub fn addresses(&self) -> Vec<Address> {
+        let mut addresses = vec![];
+        for i in 0..self.address_count() {
+            addresses.push(self.get_address(i));
+        }
+        addresses
+    }
+
+    /// Generated address count for this wallet
+    pub fn address_count(&self) -> u64 {
+        self.addrs.count
+    }
+
+    /// Transfers funds between addresses
+    pub async fn transfer(
+        &self,
+        sender: &Address,
+        rcvr: &Address,
+        amt: Dusk,
+        gas: Gas,
+    ) -> Result<Transaction, Error> {
+        if let Some(wallet) = &self.wallet {
+            // make sure we own the sender address
+            if !sender.is_owned() {
+                return Err(Error::Unauthorized);
+            }
+            // make sure amount is positive
+            if amt == 0 {
+                return Err(Error::AmountIsZero);
+            }
+            // check gas limits
+            if gas.limit < MIN_GAS_LIMIT {
+                return Err(Error::NotEnoughGas);
+            }
```

Comment:
Enjoyed reading through the thought process. To me personally, these `if`s don't hurt my eyes, but if we were to change it, I'd go with the macro!

Reply:
It doesn't hurt the eyes at all, but it's more about readability. I'll try to explain from where my thoughts come from:
those are preconditions: you're basically saying: those are sort of invariants, if those are not meet, you shouldn't even execute the rest of the code.

In rust, you usually define those invariants using `assert!`, this is also helps the compiler to optmize code. Take this, for example:
```rust
fn do_something_with_array(array: &[u8]) -> u8 {
    assert!(array.len() >= 5);
    array[0] + array[1] + array[2] + array[3] + array[4]
}
```

Basically it tells that array should always have at least five elements; if not, it should be panic because something is really wrong: this allows LLVM to realize that the later asserts are unreachable and elides them.

Now, in our case we don't do that for compiler optimization, however is always good states invariants at the beginning. In addition, we don't want to panic in case the invariant is not met, we want to recover from it. We basically want to bail sooner from the function body, we don't care about the rest of the code. That's why it would be better to "list" all the invariants together, and ensure it's clear they're invariants and not just function's logic per se: they're prerequisite to execute the function's body.

Reply:
Yeah makes sense, thanks for the thorough explanation. It is indeed not logic at all. Will explore the macro approach :)